File: composeApp/src/commonMain/kotlin/com/aus/notelikeus/data/backup/BackupAttachments.kt
package com.aus.notelikeus.data.backup

import com.aus.notelikeus.data.attachments.AttachmentLocalStorage
import com.aus.notelikeus.data.attachments.MAX_ATTACHMENT_BYTES
import com.aus.notelikeus.data.attachments.PendingAttachmentStore
import com.aus.notelikeus.data.attachments.createAttachmentId
import com.aus.notelikeus.data.attachments.pendingStoragePath
import com.aus.notelikeus.domain.model.Attachment
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

internal const val MIN_BACKUP_VERSION_WITH_ATTACHMENTS = 3
internal const val MAX_NOTE_BACKUP_ATTACHMENTS = 20

private val ALLOWED_IMAGE_MIMES = setOf(
    "image/jpeg",
    "image/jpg",
    "image/png",
    "image/webp",
    "image/gif",
)

fun interface AttachmentBackupReader {
    suspend fun readBytes(attachment: Attachment): ByteArray?
}

@OptIn(ExperimentalEncodingApi::class)
internal fun encodeAttachmentBytes(bytes: ByteArray): String = Base64.encode(bytes)

@OptIn(ExperimentalEncodingApi::class)
internal fun decodeAttachmentBytes(dataBase64: String): ByteArray? =
    runCatching { Base64.decode(dataBase64.trim()) }.getOrNull()

internal fun extensionFromMime(mimeType: String): String = when (mimeType.lowercase()) {
    "image/png" -> "png"
    "image/webp" -> "webp"
    "image/gif" -> "gif"
    else -> "jpg"
}

internal fun isAllowedBackupImageMime(mimeType: String): Boolean =
    mimeType.lowercase() in ALLOWED_IMAGE_MIMES

internal suspend fun attachmentsToBackupDtos(
    attachments: List<Attachment>,
    reader: AttachmentBackupReader?,
): List<AttachmentBackupDto> {
    if (reader == null || attachments.isEmpty()) return emptyList()
    val exported = mutableListOf<AttachmentBackupDto>()
    for (attachment in attachments) {
        if (exported.size >= MAX_NOTE_BACKUP_ATTACHMENTS) break
        if (attachment.type.isNotBlank() && !attachment.type.equals("image", ignoreCase = true)) {
            continue
        }
        val mimeType = attachment.mimeType?.takeIf { it.isNotBlank() } ?: "image/jpeg"
        if (!isAllowedBackupImageMime(mimeType)) continue
        val bytes = reader.readBytes(attachment) ?: continue
        if (bytes.isEmpty() || bytes.size > MAX_ATTACHMENT_BYTES) continue
        exported += AttachmentBackupDto(
            id = attachment.id,
            type = "image",
            mimeType = mimeType,
            sizeBytes = bytes.size.toLong(),
            dataBase64 = encodeAttachmentBytes(bytes),
            extension = extensionFromMime(mimeType),
        )
    }
    return exported
}

internal fun attachmentsFromBackupDtos(
    dtos: List<AttachmentBackupDto>,
    noteId: Long,
    backupVersion: Int,
    localStorage: AttachmentLocalStorage?,
): List<Attachment> {
    if (backupVersion < MIN_BACKUP_VERSION_WITH_ATTACHMENTS || dtos.isEmpty()) {
        return emptyList()
    }
    val restored = mutableListOf<Attachment>()
    for (dto in dtos) {
        if (restored.size >= MAX_NOTE_BACKUP_ATTACHMENTS) break
        if (dto.type.isNotBlank() && !dto.type.equals("image", ignoreCase = true)) continue
        val mimeType = dto.mimeType?.takeIf { it.isNotBlank() } ?: "image/jpeg"
        if (!isAllowedBackupImageMime(mimeType)) continue
        val bytes = decodeAttachmentBytes(dto.dataBase64) ?: continue
        if (bytes.isEmpty() || bytes.size > MAX_ATTACHMENT_BYTES) continue
        val id = dto.id?.trim()?.takeIf { it.isNotEmpty() } ?: createAttachmentId()
        val extension = dto.extension?.trim()?.takeIf { it.isNotEmpty() } ?: extensionFromMime(mimeType)
        val storagePath = localStorage?.persistImageBytes(bytes, extension)
            ?: pendingStoragePath(id).also { PendingAttachmentStore.put(id, bytes, mimeType) }
        restored += Attachment(
            id = id,
            noteId = noteId,
            storagePath = storagePath,
            type = "image",
            mimeType = mimeType,
            sizeBytes = bytes.size.toLong(),
        )
    }
    return restored
}

File: web/src/lib/backup/backupAttachments.ts
import { getAttachmentBlobStore } from '@/lib/attachments/attachmentBlobStoreRegistry';
import {
  createAttachmentId,
  isPendingAttachment,
  isR2Attachment,
  MAX_ATTACHMENT_BYTES,
  pendingStoragePath,
  ATTACHMENT_PENDING_PREFIX,
} from '@/lib/attachments/attachmentPaths';
import { peekPendingAttachment, storePendingAttachment } from '@/lib/attachments/pendingAttachmentStore';
import type { Attachment } from '@/types/attachment';
import type { Note } from '@/types/note';

export const MIN_BACKUP_VERSION_WITH_ATTACHMENTS = 3;
export const MAX_NOTE_BACKUP_ATTACHMENTS = 20;

const ALLOWED_IMAGE_MIMES = new Set([
  'image/jpeg',
  'image/jpg',
  'image/png',
  'image/webp',
  'image/gif',
]);

export interface BackupAttachmentDto {
  id?: string;
  type?: string;
  mimeType?: string;
  sizeBytes?: number;
  dataBase64: string;
  extension?: string;
}

export function extensionFromMime(mimeType: string): string {
  switch (mimeType.toLowerCase()) {
    case 'image/png':
      return 'png';
    case 'image/webp':
      return 'webp';
    case 'image/gif':
      return 'gif';
    default:
      return 'jpg';
  }
}

export function isAllowedBackupImageMime(mimeType: string): boolean {
  return ALLOWED_IMAGE_MIMES.has(mimeType.toLowerCase());
}

export function bytesToBase64(bytes: Uint8Array): string {
  let binary = '';
  const chunk = 0x8000;
  for (let i = 0; i < bytes.length; i += chunk) {
    binary += String.fromCharCode(...bytes.subarray(i, i + chunk));
  }
  return btoa(binary);
}

export function base64ToBytes(dataBase64: string): Uint8Array | null {
  try {
    const binary = atob(dataBase64.trim());
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i += 1) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  } catch {
    return null;
  }
}

async function readAttachmentBlob(note: Note, attachment: Attachment): Promise<Blob | null> {
  if (isPendingAttachment(attachment.storagePath)) {
    const pendingId = attachment.storagePath.slice(ATTACHMENT_PENDING_PREFIX.length);
    return peekPendingAttachment(pendingId)?.blob ?? null;
  }
  if (isR2Attachment(attachment.storagePath)) {
    try {
      return await getAttachmentBlobStore().download(note.id, attachment.id);
    } catch {
      return null;
    }
  }
  return null;
}

export async function attachmentsToBackupDtos(note: Note): Promise<BackupAttachmentDto[]> {
  const exported: BackupAttachmentDto[] = [];
  for (const attachment of note.attachments) {
    if (exported.length >= MAX_NOTE_BACKUP_ATTACHMENTS) break;
    if (attachment.type && attachment.type !== 'image') continue;
    const mimeType = attachment.mimeType?.trim() || 'image/jpeg';
    if (!isAllowedBackupImageMime(mimeType)) continue;
    const blob = await readAttachmentBlob(note, attachment);
    if (!blob || blob.size === 0 || blob.size > MAX_ATTACHMENT_BYTES) continue;
    const bytes = new Uint8Array(await blob.arrayBuffer());
    exported.push({
      id: attachment.id,
      type: 'image',
      mimeType,
      sizeBytes: bytes.byteLength,
      dataBase64: bytesToBase64(bytes),
      extension: extensionFromMime(mimeType),
    });
  }
  return exported;
}

export function attachmentsFromBackupDtos(
  dtos: unknown,
  noteId: number,
  backupVersion: number,
): Attachment[] {
  if (backupVersion < MIN_BACKUP_VERSION_WITH_ATTACHMENTS || !Array.isArray(dtos)) {
    return [];
  }
  const restored: Attachment[] = [];
  for (const raw of dtos) {
    if (restored.length >= MAX_NOTE_BACKUP_ATTACHMENTS) break;
    if (!raw || typeof raw !== 'object') continue;
    const dto = raw as BackupAttachmentDto;
    if (typeof dto.dataBase64 !== 'string') continue;
    if (dto.type && dto.type !== 'image') continue;
    const mimeType = dto.mimeType?.trim() || 'image/jpeg';
    if (!isAllowedBackupImageMime(mimeType)) continue;
    const bytes = base64ToBytes(dto.dataBase64);
    if (!bytes || bytes.byteLength === 0 || bytes.byteLength > MAX_ATTACHMENT_BYTES) continue;
    const id = dto.id?.trim() || createAttachmentId();
    const copy = new Uint8Array(bytes.byteLength);
    copy.set(bytes);
    const blob = new Blob([copy], { type: mimeType });
    storePendingAttachment(id, blob, mimeType);
    restored.push({
      id,
      noteId,
      storagePath: pendingStoragePath(id),
      type: 'image',
      mimeType,
      sizeBytes: bytes.byteLength,
    });
  }
  return restored;
}
